File: 2020/javascript/day6/answer.js
const fs = require("fs");

const parse = (input) =>
    input.split("\n\n").map((group) => group.trim().split("\n"));

const groups = parse(fs.readFileSync("input", { encoding: "utf8", flag: "r" }));

function countUnique(groups) {
    return groups
        .map((group) => {
            set = new Set(group.join(""));
            return set.size;
        })
        .reduce((total, current) => total + current, 0);
}

function countSame(groups) {
    return groups
        .map((group) =>
            group
                .map((line) => new Set(line))
                .reduce(
                    (intersectionSet, set) =>
                        new Set([...intersectionSet].filter((x) => set.has(x))),
                    new Set(group[0])
                )
        )
        .map((set) => set.size)
        .reduce((a, b) => a + b, 0);
}

console.log(`Part 1: ${countUnique(groups)}`);
console.log(`Part 2: ${countSame(groups)}`);

File: 2020/python/day6/answer.py
import re


def parse(input):
    return input.replace('\n\n', ',').replace('\n', ' ').split(',')


def countUnique(group):
    return len(set(''.join(group)))


def countSame(group):
    return len(set.intersection(*[set(i) for i in group]))


with open('input') as f:
    input = parse(f.read())
    choices = [i.strip().split(' ') for i in input]

    print(f'Part 1: {sum([countUnique(i) for i in choices])}')
    print(f'Part 2: {sum([countSame(i) for i in choices])}')
